core/src/main/scala/org/apache/spark/rdd/CoalescedRDD.scala

```diff
@@ -337,19 +353,23 @@ private class PartitionCoalescer(maxPartitions: Int, prev: RDD[_], balanceSlack:
    * load balanced and grouped by locality
    * @return array of partition groups
    */
-  def run(): Array[PartitionGroup] = {
-    setupGroups(math.min(prev.partitions.length, maxPartitions))   // setup the groups (bins)
-    throwBalls() // assign partitions (balls) to each group (bins)
+  def coalesce(maxPartitions: Int, prev: RDD[_]): Array[PartitionGroup] = {
+    setupGroups(math.min(prev.partitions.length, maxPartitions), prev)   // setup the groups (bins)
+    throwBalls(maxPartitions, prev, balanceSlack) // assign partitions (balls) to each group (bins)
     getPartitions
   }
 }
 
-private case class PartitionGroup(prefLoc: Option[String] = None) {
+/**
+ * A group of [[Partition]]s
+ * @param prefLoc preferred location for the partition group
+ */
+case class PartitionGroup(prefLoc: Option[String] = None) {
```

Comment:
1. move this closer to PartitionCoalescer
2. make it a normal class (case class is difficult to maintain backward compatibility due the pattern matching)